File: LC100/lc56.py
"""
Merge Interval
Given a collection of intervals, merge all overlapping intervals.

For example,
Given [1,3],[2,6],[8,10],[15,18],
return [1,6],[8,10],[15,18]
"""
# Definition for an interval.
# class Interval(object):
#     def __init__(self, s=0, e=0):
#         self.start = s
#         self.end = e


def merge(intervals):
    output = []
    intervals.sort(key=lambda i: i.start)
    for an_interval in intervals:
        if output and an_interval.start <= output[-1].end:
            output[-1].end = max(an_interval.end, output[-1].end)
        else:
            output.append(intervals)

    return

File: LC100/lc56.js
class interval {
    constructor(start=0, end=0) {
        this.start = start;
        this.end = end;
    }
}

function merge(intervals) {
    var output = [];
    intervals.sort( (a, b) => {
        return a.start - b.start;
    });

    var last;
    for (var an_interval of intervals) {
        if (output.length > 0 && an_interval.start <= last.end) {
            last.end = Math.max(an_interval.end, last.end);
        } else {
            output.push(an_interval);
            last = an_interval;
        }
    }
    return output;
}


var intervals = [
    new interval(1, 3),
    new interval(2, 6),
    new interval(8, 10),
    new interval(15, 18),
];
merge(intervals);
